example/csv/src/main/java/org/apache/calcite/adapter/csv/JsonEnumerator.java

```diff
@@ -67,15 +67,8 @@ static JsonDataConverter deduceRowType(RelDataTypeFactory typeFactory, Source so
       objectMapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true)
           .configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
           .configure(JsonParser.Feature.ALLOW_COMMENTS, true);
-      if (source.file().exists() && source.file().length() > 0) {
-        if ("file".equals(source.protocol())) {
-          //noinspection unchecked
-          jsonObj = objectMapper.readValue(source.file(), Object.class);
-        } else {
-          //noinspection unchecked
-          jsonObj = objectMapper.readValue(source.url(), Object.class);
-        }
-      }
+
+      jsonObj = objectMapper.readValue(source.reader(), Object.class);
```

Comment:
Technically speaking, this makes error messages harder to understand. For instance, `objectMapper.readValue(File,...` can produce targeted exception messages and refer to the file and the line number. However, if you abstract that behind `reader`, then targeted exceptions are not possible

Reply:
yes. that's a fair point. Let me adjust it